### ISSUE TITLE
RKE2 Additional Server failed to join the RKE2 HA cluster

### ISSUE THREAD

**Issue by ab-mohamed** [rancher/rke2#1760]
**Environmental Info:**
RKE2 Version: 
```
rke2-server1:~ # rke2 -v
rke2 version v1.21.4+rke2r2 (d5746a33b001dcf1a24cd3256e9f1df2a43ff3dd)
go version go1.16.6b7
```

Node(s) CPU architecture, OS, and Version: 
* Three-node HA cluster
* OS: SLES15 SP3
* Cloud Provider: GCP
* Instance image: `n1-standard-2`

Cluster Configuration:
I am deploying an RKE2 HA cluster on GCP with three RKE2 servers and two RKE2 agents. I am following the instructions in RKE2 documentation HA deployment (https://docs.rke2.io/install/ha/)

**Describe the bug:**
The second RKE2 server was not able to join the cluster.

**Steps To Reproduce:**
1. `rke2-server.ab.lab` is my first RKE2 Server, and it is working fine:
```
rke2-server1:~> sudo /var/lib/rancher/rke2/bin/kubectl get nodes -o wide --kubeconfig=/tmp/config 
NAME           STATUS   ROLES                       AGE    VERSION          INTERNAL-IP    EXTERNAL-IP   OS-IMAGE                              KERNEL-VERSION         CONTAINER-RUNTIME
rke2-server1   Ready    control-plane,etcd,master   123m   v1.21.4+rke2r2   192.168.0.11   <none>        SUSE Linux Enterprise Server 15 SP3   5.3.18-59.19-default   containerd://1.4.8-k3s1
```

I use `rke2.ab.lab` in my KubeConfig file:
```
ab@rke2-server1:~> sudo cat /tmp/config
[OUTPUT TRIMMED]
    server: https://rke2.ab.lab:6443
[OUTPUT TRIMMED]
```

Here are the `config.yaml` file contents:
```
rke2-server1:~> sudo cat /etc/rancher/rke2/config.yaml 
tls-san:
  - rke2.ab.lab
```
Where `rke2.ab.lab` is the frontend IP of an internal GCP TCP load balancer. The load balancer permits ports `TCP|6443` and `TCP|9345`, and it uses port `TCP|6443` for the health check.

2. Here are the RKE2 server2 `config.yaml` file contents:
```
rke2-server2:~> sudo cat /etc/rancher/rke2/config.yaml
server: https://rke2.ab.lab:9345
token:  <TOKEN VALUE>
tls-san:
  - rke2.ab.lab
```

3. Lunch the RKE2 installer on `rke2-server2`:
```
rke2-server2:~ # curl -sfL https://get.rke2.io | sh -
[INFO]  finding release for channel stable
[INFO]  using v1.21.4+rke2r2 as release
[INFO]  downloading checksums at https://github.com/rancher/rke2/releases/download/v1.21.4+rke2r2/sha256sum-amd64.txt
[INFO]  downloading tarball at https://github.com/rancher/rke2/releases/download/v1.21.4+rke2r2/rke2.linux-amd64.tar.gz
[INFO]  verifying tarball
[INFO]  unpacking tarball file to /usr/local
```

4. Enable the `rke2-server` systemd service on `rke2-server2`:
```
rke2-server2:~ # systemctl enable --now rke2-server.service
Created symlink /etc/systemd/system/multi-user.target.wants/rke2-server.service → /usr/local/lib/systemd/system/rke2-server.service.
Job for rke2-server.service failed because the control process exited with error code.
See "systemctl status rke2-server.service" and "journalctl -xe" for details.
```

5. Logs shows the following error messages:
```
Sep 06 17:13:43 rke2-server2 systemd[1]: rke2-server.service: Main process exited, code=exited, status=1/FAILURE
Sep 06 17:13:43 rke2-server2 systemd[1]: rke2-server.service: Failed with result 'exit-code'.
Sep 06 17:13:43 rke2-server2 systemd[1]: Failed to start Rancher Kubernetes Engine v2 (server).
Sep 06 17:13:49 rke2-server2 systemd[1]: rke2-server.service: Scheduled restart job, restart counter is at 64.
Sep 06 17:13:49 rke2-server2 systemd[1]: Stopped Rancher Kubernetes Engine v2 (server).
Sep 06 17:13:49 rke2-server2 systemd[1]: Starting Rancher Kubernetes Engine v2 (server)...
Sep 06 17:13:49 rke2-server2 sh[4133]: + /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service
Sep 06 17:13:49 rke2-server2 sh[4135]: Failed to get unit file state for nm-cloud-setup.service: No such file or directory
Sep 06 17:13:49 rke2-server2 rke2[4151]: time="2021-09-06T17:13:49Z" level=warning msg="not running in CIS mode"
Sep 06 17:13:49 rke2-server2 rke2[4151]: time="2021-09-06T17:13:49Z" level=info msg="Starting rke2 v1.21.4+rke2r2 (d5746a33b001dcf1a24cd3256e9f1df2a43ff3dd)"
Sep 06 17:13:49 rke2-server2 rke2[4151]: time="2021-09-06T17:13:49Z" level=fatal msg="starting kubernetes: preparing server: failed to get CA certs: Get \"https://rke2.ab.lab:9345/cacerts\": dial tcp 192.168.0.100:9345: connect: connection refused"
````

6.  CURL is not able to establish the secure connection as shown below:

`rke2-server1`:
```
rke2-server1:~> curl https://rke2.ab.lab:9345/cacerts
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

But it  works with the  `--insecure` flag:
```
rke2-server1:~> curl --insecure https://rke2.ab.lab:9345/cacerts
-----BEGIN CERTIFICATE-----
[OUTPUT TRIMMED]
-----END CERTIFICATE----
```

From `rke2-server2`:
```
rke2-server2:~> curl https://rke2.ab.lab:9345/cacerts
curl: (7) Failed to connect to rke2.ab.lab port 9345: Connection refused
```
```
rke2-server2:~> curl --insecure https://rke2.ab.lab:9345/cacerts
curl: (7) Failed to connect to rke2.ab.lab port 9345: Connection refused
```
```
rke2-server2:~> curl https://rke2-server1.ab.lab:9345/cacerts
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html
curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```
```
rke2-server2:~> curl --insecure https://rke2-server1.ab.lab:9345/cacerts
-----BEGIN CERTIFICATE-----
[OUTPUT TRIMMED]
-----END CERTIFICATE-----
````

7. Reference to https://github.com/rancher/rke2/issues/518, I see the `tls-san` that I created (rke2.ab.lab):
```
rke2-server1:/var/lib/rancher/rke2/bin> sudo ./kubectl --kubeconfig=/tmp/config get secret -o yaml rke2-serving -n kube-system | grep -B12 -A9 rke2.ab.lab
metadata:
  annotations:
    listener.cattle.io/cn-10.43.0.1: 10.43.0.1
    listener.cattle.io/cn-127.0.0.1: 127.0.0.1
    listener.cattle.io/cn-192.168.0.11: 192.168.0.11
    listener.cattle.io/cn-192.168.0.100: 192.168.0.100
    listener.cattle.io/cn-kubernetes: kubernetes
    listener.cattle.io/cn-kubernetes.default: kubernetes.default
    listener.cattle.io/cn-kubernetes.default.svc: kubernetes.default.svc
    listener.cattle.io/cn-kubernetes.default.svc.cluster.local: kubernetes.default.svc.cluster.local
    listener.cattle.io/cn-localhost: localhost
    listener.cattle.io/cn-rke2-server1.ab.lab: rke2-server1.ab.lab
    listener.cattle.io/cn-rke2.ab.lab: rke2.ab.lab
    listener.cattle.io/fingerprint: SHA1=BA3092E21E063A994E66F31CC42FDA661FD685B3
  creationTimestamp: "2021-09-06T15:47:08Z"
  name: rke2-serving
  namespace: kube-system
  resourceVersion: "19702"
  uid: 3222b10e-02c6-4f25-add5-c07c63ff6d77
type: kubernetes.io/tls
```

Would you please advise how to fix this issue?
 
**Expected behavior:**
The second RKE2 server joins the cluster

**Actual behavior:**
The second RKE2 sever failed to join the cluster, as mentioned in the above sections.

**Comment by ab-mohamed** [comment]
I rebuild the environment. This time I did not install any RKE2 servers. First, I ensured that the GCP load balancer health check works fine by  simulating opened `TCP|6443` port on the three nodes using the following commands:
```
rke2-server1:~ # timeout 60s socat - TCP-LISTEN:6443,fork
rke2-server2:~ # timeout 60s socat - TCP-LISTEN:6443,fork
rke2-server3:~ # timeout 60s socat - TCP-LISTEN:6443,fork
```

Then I check the load balancer backend nodes' health status. All nodes status were `HEALTHY`:
```
$ gcloud compute backend-services get-health rke2-ilb --region us-west1
---
backend: https://www.googleapis.com/compute/v1/projects/rancher-demo-101/zones/us-west1-a/instanceGroups/rke2-ig1
status:
  healthStatus:
  - forwardingRule: https://www.googleapis.com/compute/v1/projects/rancher-demo-101/regions/us-west1/forwardingRules/rke2-frontend1
    forwardingRuleIp: 192.168.0.100
    healthState: HEALTHY
    instance: https://www.googleapis.com/compute/v1/projects/rancher-demo-101/zones/us-west1-a/instances/rke2-server1
    ipAddress: 192.168.0.11
    port: 80
  kind: compute#backendServiceGroupHealth
---
backend: https://www.googleapis.com/compute/v1/projects/rancher-demo-101/zones/us-west1-b/instanceGroups/rke2-ig2
status:
  healthStatus:
  - forwardingRule: https://www.googleapis.com/compute/v1/projects/rancher-demo-101/regions/us-west1/forwardingRules/rke2-frontend1
    forwardingRuleIp: 192.168.0.100
    healthState: HEALTHY
    instance: https://www.googleapis.com/compute/v1/projects/rancher-demo-101/zones/us-west1-b/instances/rke2-server2
    ipAddress: 192.168.0.12
    port: 80
  kind: compute#backendServiceGroupHealth
---
backend: https://www.googleapis.com/compute/v1/projects/rancher-demo-101/zones/us-west1-c/instanceGroups/rke2-ig3
status:
  healthStatus:
  - forwardingRule: https://www.googleapis.com/compute/v1/projects/rancher-demo-101/regions/us-west1/forwardingRules/rke2-frontend1
    forwardingRuleIp: 192.168.0.100
    healthState: HEALTHY
    instance: https://www.googleapis.com/compute/v1/projects/rancher-demo-101/zones/us-west1-c/instances/rke2-server3
    ipAddress: 192.168.0.13
    port: 80
  kind: compute#backendServiceGroupHealth
```

I installed the RKE2 server on `rke2-server1` node again. The installation process was completed successfully, but `rke2-server2` could not join the cluster again:
```
rke2-server2:~> curl --insecure https://rke2.ab.lab:9345/cacerts
curl: (7) Failed to connect to rke2.ab.lab port 9345: Connection refused
```
```
ab@rke2-server2:~> curl --insecure https://rke2-server1.ab.lab:9345/cacerts
-----BEGIN CERTIFICATE-----
[OUTPUT TRIMMED]
-----END CERTIFICATE-----
```

**Comment by brandond** [comment]
You said that `rke2.ab.lab` is a GCP TCP load balancer, but clearly it's not accepting connections on the port. Not even curl can make a connection to it. Are you sure you've got the load-balancer configured correctly?

```console
rke2-server2:~> curl --insecure https://rke2.ab.lab:9345/cacerts
curl: (7) Failed to connect to rke2.ab.lab port 9345: Connection refused
```

**Comment by ab-mohamed** [comment]
@brandond, 

Thank you for your reply. 

There are three types of Google Cloud load balancers:
* HTTP(S) load balancers (Layer7)
* TCP load balancers (Layer4)
* UDP load balancer (Layer4)

I use the TCP load balancer with three unmanaged instance groups `rke2-ig1`, `rke2-ig2`, and `rke2-ig3`. Each instance group has one RKE2 server node. 
The three RKE2 server nodes are created in different availability zones within the same region, `us-west1`. 

The health check rule function was tested successfully, as shared in my previous comment. 

Any advice?

**Comment by brandond** [comment]
The external load balancer needs to be curl-able. If you get a connection reset from it then either something is wrong on that side, or you've got something in between your node and the LB that is denying your connection.

**Comment by galal-hussein** [comment]
After debugging, I think GCP is mishandling the routing between nodes if you request the load balancer from the same Instance group, as a workaround you can only add the first node as a backend and after joining the other two nodes, adds them as backends to the load balancer:

- setup the load balancer with only one backend (the first server)
- start rke2 on the first server
- start rke2 on the second and third server
- add the second and third instance groups to the load balancer backends

**Comment by brandond** [comment]
Is it not health-checking properly, and trying to send traffic to nodes that aren't up yet?

**Comment by galal-hussein** [comment]
> Is it not health-checking properly, and trying to send traffic to nodes that aren't up yet?

no its not that, apparently when the second and third node try to hit the load balancer, they get routed to the localhost, instead of the working node (node1), I am still going through the configuration, but I dont see anything wrong so far

**Comment by galal-hussein** [comment]
Update: I did some more digging, and it turned out that all instances in instance group are using a floating ip that redirects them to the localhost:

```
root@hgalal-rke2-1b:~# curl https://10.240.0.121:9345
```
and then dumping traffic, as you can see its not really getting out of the server
```
root@hgalal-rke2-1b:~# tcpdump -i any dst 10.240.0.121 and dst port 9345
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked v1), capture size 262144 bytes
00:48:30.925025 IP 10.240.0.121.40980 > 10.240.0.121.9345: Flags [S], seq 1583367938, win 65495, options [mss 65495,sackOK,TS val 982902574 ecr 0,nop,wscale 7], length 0
00:48:30.925062 IP 10.240.0.121.40980 > 10.240.0.121.9345: Flags [.], ack 199090328, win 512, options [nop,nop,TS val 982902574 ecr 982902574], length 0
00:48:30.934583 IP 10.240.0.121.40980 > 10.240.0.121.9345: Flags [P.], seq 0:517, ack 1, win 512, options [nop,nop,TS val 982902584 ecr 982902574], length 517
00:48:30.935250 IP 10.240.0.121.40980 > 10.240.0.121.9345: Flags [.], ack 991, win 505, options [nop,nop,TS val 982902585 ecr 982902584], length 0
00:48:30.935626 IP 10.240.0.121.40980 > 10.240.0.121.9345: Flags [P.], seq 517:524, ack 991, win 505, options [nop,nop,TS val 982902585 ecr 982902584], length 7
00:48:30.935757 IP 10.240.0.121.40980 > 10.240.0.121.9345: Flags [.], ack 1015, win 505, options [nop,nop,TS val 982902585 ecr 982902585], length 0
00:48:30.936684 IP 10.240.0.121.40980 > 10.240.0.121.9345: Flags [R.], seq 524, ack 1016, win 512, options [nop,nop,TS val 982902586 ecr 982902586], length 0
```

After removing the node from the instance group, you can see that its getting out from the server's NIC:
```
root@hgalal-rke2-1b:~# tcpdump  -i any dst 10.240.0.121 and dst port 9345
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked v1), capture size 262144 bytes
00:55:29.789974 IP hgalal-rke2-1b.c.rancher-dev.internal.36296 > 10.240.0.121.9345: Flags [S], seq 4083947126, win 65320, options [mss 1420,sackOK,TS val 4189886213 ecr 0,nop,wscale 7], length 0
00:55:29.791781 IP hgalal-rke2-1b.c.rancher-dev.internal.36296 > 10.240.0.121.9345: Flags [.], ack 3746415633, win 511, options [nop,nop,TS val 4189886215 ecr 3530264310], length 0
00:55:29.801289 IP hgalal-rke2-1b.c.rancher-dev.internal.36296 > 10.240.0.121.9345: Flags [P.], seq 0:517, ack 1, win 511, options [nop,nop,TS val 4189886225 ecr 3530264310], length 517
00:55:29.802064 IP hgalal-rke2-1b.c.rancher-dev.internal.36296 > 10.240.0.121.9345: Flags [.], ack 990, win 504, options [nop,nop,TS val 4189886225 ecr 3530264320], length 0
00:55:29.802626 IP hgalal-rke2-1b.c.rancher-dev.internal.36296 > 10.240.0.121.9345: Flags [P.], seq 517:524, ack 990, win 504, options [nop,nop,TS val 4189886226 ecr 3530264320], length 7
00:55:29.802844 IP hgalal-rke2-1b.c.rancher-dev.internal.36296 > 10.240.0.121.9345: Flags [.], ack 1014, win 504, options [nop,nop,TS val 4189886226 ecr 3530264321], length 0
00:55:29.803401 IP hgalal-rke2-1b.c.rancher-dev.internal.36296 > 10.240.0.121.9345: Flags [R.], seq 524, ack 1015, win 504, options [nop,nop,TS val 4189886227 ecr 3530264322], length 0

```

**Comment by brandond** [comment]
So it's some weird cloud-provider specific behavior where they push the load balancer traffic routing down to the node?

**Comment by galal-hussein** [comment]
> So it's some weird cloud-provider specific behavior where they push the load balancer traffic routing down to the node?

Exactly

**Comment by ab-mohamed** [comment]
@galal-hussein and @brandond: Thank you for your help. 

Digging more in GCP documentation. It seems that, by default, requests to the load balancer that originate from any of the backend VMs -the server VMs being load balanced- are always answered by the same VM making the request.
A request is always sent to the VM that makes the request, and health check information is ignored. Further, the backend can respond to traffic sent using protocols and destination ports other than those configured on the load balancer's internal forwarding rule.

@galal-hussein: your workaround works. I made a slight change to ensure that the HA function works by executing the following steps:

1. Create three instance groups:
 * `rke2-ig1` that contains `rke2-server1` node.
 * `rke2-ig2` that contains `rke2-server2` node. 
 *  `rke2-ig3` that contains `rke2-server3` node.
2. Setup the load balancer with only one backend (rke2-ig1).
2. Start RKE2 server on `rke2-server1`node.
3. Start RKE2 agent on `rke2-agent1` node.
3. Start RKE2 server on `rke2-server2`. 
4. Add `rke2-ig2` instance group to the load balancer backend.
5. Start RKE2 server on `rke2-server3`.
6. Add `rke2-ig3` instance group to the load balancer backend.
7. Power off `rke2-server1` node. So we have two RKE2 servers now (`rke2-server2` and `rke2-server3`).
8. Start RKE2 agent on `rke2-agent2` node.

The results look fine:
```
ab@rke2-bastion-host:~> kubectl cluster-info
Kubernetes master is running at https://rke2.ab.lab:6443
CoreDNS is running at https://rke2.ab.lab:6443/api/v1/namespaces/kube-system/services/rke2-coredns-rke2-coredns:udp-53/proxy
```
```
ab@rke2-bastion-host:~> kubectl get nodes -o wide 
NAME           STATUS   ROLES                       AGE    VERSION          INTERNAL-IP    EXTERNAL-IP   OS-IMAGE                              KERNEL-VERSION         CONTAINER-RUNTIME
rke2-agent1    Ready    <none>                      107m   v1.21.4+rke2r3   192.168.0.21   <none>        SUSE Linux Enterprise Server 15 SP3   5.3.18-59.19-default   containerd://1.4.8-k3s1
rke2-agent2    Ready    <none>                      82m    v1.21.4+rke2r3   192.168.0.22   <none>        SUSE Linux Enterprise Server 15 SP3   5.3.18-59.19-default   containerd://1.4.8-k3s1
rke2-server1   Ready    control-plane,etcd,master   115m   v1.21.4+rke2r3   192.168.0.11   <none>        SUSE Linux Enterprise Server 15 SP3   5.3.18-59.19-default   containerd://1.4.8-k3s1
rke2-server2   Ready    control-plane,etcd,master   97m    v1.21.4+rke2r3   192.168.0.12   <none>        SUSE Linux Enterprise Server 15 SP3   5.3.18-59.19-default   containerd://1.4.8-k3s1
rke2-server3   Ready    control-plane,etcd,master   89m    v1.21.4+rke2r3   192.168.0.13   <none>        SUSE Linux Enterprise Server 15 SP3   5.3.18-59.19-default   containerd://1.4.8-k3s1
```

I am looking now for a permanent resolution instead of the mentioned workaround.

**Comment by ab-mohamed** [comment]
Hi, 

Based on [Sending requests from load balanced VMs](https://cloud.google.com/load-balancing/docs/internal/setting-up-internal#test-from-backend-vms):

> Don't rely on making requests to an internal TCP/UDP load balancer from a VM being load balanced (in the backend service for that load balancer). A request is always sent to the VM that makes the request, and health check information is ignored. Further, the backend can respond to traffic sent using protocols and destination ports other than those configured on the load balancer's internal forwarding rule.
So, I now understand the reason behind the load balancer behavior. 

I have possible two resolutions for the above-mentioned load balancer behavior, but I need to ask for your help (@galal-hussein and @brandond) to answer the following question:

When `rke2-server2` joins the RKE2 cluster for the first time, it uses the Load Balancer frontend IP domain name (`rke2.ab.lab` in my case). Is it a one-time action? Or is it a must during every OS reboot? 
I mean, when the `rke2-server2` node is rebooted, does it register again against the `rke2.ab.lab` domain?

**Comment by ab-mohamed** [comment]
Another question. Is the communication between the three RKE2 server nodes, `rke2-server1`, `rke2-server2`, and `rke2-server3` based on the load balancer front end IP?

**Comment by brandond** [comment]
> Another question. Is the communication between the three RKE2 server nodes, `rke2-server1`, `rke2-server2`, and `rke2-server3` based on the load balancer front end IP?

https://docs.rke2.io/install/ha/
> Agents register through the fixed registration address. However, when RKE2 launches the kubelet and it must connect to the Kubernetes api-server, it does so through the rke2 agent process, which acts as a client-side load balancer.

This is also true of servers registering to join the cluster. The initial connection uses the fixed registration address, but once the node has joined the cluster all communication is peer to peer.

**Comment by ab-mohamed** [comment]
@brandond Thank you for your update. 

Please allow me to summarize what I understand here:
1. RKE2 servers or agents *initial* registration (based on the `/etc/rancher/rke2/config.yaml` config file and starting the `rke2-server` server or `rke2-agent` systemd services) use a fixed address (in my case, it is  `rke2.ab.lab` which is the load balancer frontend IP address)
2. Once the initial registration is done, any further communications between RKE2 servers or agents nodes do *NOT* use the fixed address `rke2.ab.lab`.

**Comment by ab-mohamed** [comment]
@brandond and @galal-hussein: Any updates regarding my last comment? 

Thank you. :)

**Comment by AmreeshTyagi** [comment]
I have still observed the same issue.
```sh
[INFO]  using v1.24.3-rc3+rke2r1 as release
[INFO]  downloading checksums at https://github.com/rancher/rke2/releases/download/v1.24.3-rc3+rke2r1/sha256sum-amd64.txt
[INFO]  downloading tarball at https://github.com/rancher/rke2/releases/download/v1.24.3-rc3+rke2r1/rke2.linux-amd64.tar.gz
[INFO]  verifying tarball
[INFO]  unpacking tarball file to /usr/local
Created symlink /etc/systemd/system/multi-user.target.wants/rke2-server.service → /usr/local/lib/systemd/system/rke2-server.service.
Job for rke2-server.service failed because the control process exited with error code.
See "systemctl status rke2-server.service" and "journalctl -xe" for details.
```

This is what I got from journelctl:

```sh
root@app-cluster-1-master-2:~# journalctl -xe
Jul 20 12:20:29 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:29Z" level=info msg="Waiting for API server to become available"
Jul 20 12:20:30 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:30Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:20:35 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:35Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:20:39 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:20:39.723Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:20:39 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:39Z" level=info msg="Failed to test data store connection: context deadline exceeded"
Jul 20 12:20:40 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:20:40.874Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:20:40 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:40Z" level=error msg="Failed to check local etcd status for learner management: context de>
Jul 20 12:20:40 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:40Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:20:45 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:45Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:20:50 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:50Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:20:54 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:20:54.725Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:20:54 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:54Z" level=info msg="Failed to test data store connection: context deadline exceeded"
Jul 20 12:20:55 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:20:55.873Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:20:55 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:55Z" level=error msg="Failed to check local etcd status for learner management: context de>
Jul 20 12:20:55 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:55Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:20:55 app-cluster-1-master-2 sshd[9127]: Invalid user vyos from 92.81.58.2 port 39794
Jul 20 12:20:55 app-cluster-1-master-2 sshd[9127]: pam_unix(sshd:auth): check pass; user unknown
Jul 20 12:20:55 app-cluster-1-master-2 sshd[9127]: pam_unix(sshd:auth): authentication failure; logname= uid=0 euid=0 tty=ssh ruser= rhost=92.81.58.2
Jul 20 12:20:58 app-cluster-1-master-2 sshd[9127]: Failed password for invalid user vyos from 92.81.58.2 port 39794 ssh2
Jul 20 12:20:58 app-cluster-1-master-2 sshd[9127]: Received disconnect from 92.81.58.2 port 39794:11: Bye Bye [preauth]
Jul 20 12:20:58 app-cluster-1-master-2 sshd[9127]: Disconnected from invalid user vyos 92.81.58.2 port 39794 [preauth]
Jul 20 12:20:59 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:59Z" level=info msg="Waiting for etcd server to become available"
Jul 20 12:20:59 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:20:59Z" level=info msg="Waiting for API server to become available"
Jul 20 12:21:00 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:00Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:21:05 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:05Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:21:09 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:21:09.726Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:21:09 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:09Z" level=info msg="Failed to test data store connection: context deadline exceeded"
Jul 20 12:21:10 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:21:10.873Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:21:10 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:10Z" level=error msg="Failed to check local etcd status for learner management: context de>
Jul 20 12:21:10 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:10Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:21:15 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:15Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:21:20 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:20Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:21:24 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:21:24.728Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:21:24 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:24Z" level=info msg="Failed to test data store connection: context deadline exceeded"
Jul 20 12:21:25 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:21:25.875Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:21:25 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:25Z" level=error msg="Failed to check local etcd status for learner management: context de>
Jul 20 12:21:25 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:25Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:21:29 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:29Z" level=info msg="Waiting for etcd server to become available"
Jul 20 12:21:29 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:29Z" level=info msg="Waiting for API server to become available"
Jul 20 12:21:30 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:30Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:21:35 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:35Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:21:39 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:21:39.730Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:21:39 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:39Z" level=info msg="Failed to test data store connection: context deadline exceeded"
Jul 20 12:21:40 app-cluster-1-master-2 rke2[8540]: {"level":"warn","ts":"2022-07-20T12:21:40.874Z","logger":"etcd-client","caller":"v3@v3.5.4-k3s1/retry_interceptor>
Jul 20 12:21:40 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:40Z" level=error msg="Failed to check local etcd status for learner management: context de>
Jul 20 12:21:40 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:40Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht>
Jul 20 12:21:45 app-cluster-1-master-2 rke2[8540]: time="2022-07-20T12:21:45Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: ht
```

This issue is intermittent. Not sure about reproducing steps.
Any idea? How it can be avoided? 

For now, I am just updating my terraform node count & redeploy sever sometime to make it successful.

**Comment by brandond** [comment]
@AmreeshTyagi please open an new issue. As you can see from the comments, this one ended up being caused by GCE load balancer behavior.
